### PR TITLE
vllm-ascend support chunked prefill

### PR DESCRIPTION
**Pull request by fems14**
### What this PR does / why we need it?
vllm-ascend support chunked prefill for MLA

### CODE REVIEW

File: vllm_ascend/attention/mla_v1.py [hunk 1]
```diff
@@ -358,6 +398,41 @@ def build(
...
 
+            context_lens_cpu = self.runner.input_batch.num_computed_tokens_cpu_tensor[
+                reqs_start:num_reqs]
+            max_context_len_cpu = context_lens_cpu.max().item()
+            num_prefills_with_context_cpu = (context_lens_cpu > 0).sum().item()
+            if self.chunked_prefill_enabled and max_context_len_cpu > 0:
+                max_context_chunk = (self.chunked_prefill_workspace_size //
+                                     num_prefills_with_context_cpu)
+                max_context_chunk = round_down(
+                    max_context_chunk, self.block_size)  #待确认是否需要是block_size的倍数
```
Comment (wangxiyuan) [inline vllm_ascend/attention/mla_v1.py]: do not use chinese in code

File: vllm_ascend/ascend_config.py [hunk 2]
```diff
@@ -39,6 +39,7 @@ def __init__(self, vllm_config):
         self.expert_tensor_parallel_size = int(
             additional_config.get("expert_tensor_parallel_size", 0))
         self.expert_map_path = additional_config.get("expert_map_path", None)
+        self.new_chunked = additional_config.get("new_chunked", False)
```
Comment (wangxiyuan) [inline vllm_ascend/ascend_config.py]: rename to something like `chunked_prefill_for_mla`. And the doc for `additional_config` should be updated as well

Comment (ganyi1996ppo) [reply vllm_ascend/ascend_config.py]: vLLM use chunked prefill to schedule the request by default, maybe we should remove this.

File: vllm_ascend/worker/model_runner_v1.py [hunk 3]
```diff
@@ -861,7 +866,7 @@ def _process_reqs(
         elif np.all(num_valid_tokens == 1):
             attn_state = AscendAttentionState.SpecDecoding
         # splitfuse
-        elif not ascend_config.ascend_scheduler_config.enabled or self.chunked_prefill_enabled:
+        elif self.chunked_prefill_enabled:
```
Comment (ganyi1996ppo) [inline vllm_ascend/worker/model_runner_v1.py]: This part of code still have problems, my suggest is remain unchange over this part and remove the `flash_attention` and `vanilla_mla` path at mla_backend.

File: vllm_ascend/attention/mla_v1.py [hunk 4]
```diff
@@ -582,19 +735,29 @@ def _forward_prefill(
...
-                                      self.num_heads * self.v_head_dim,
-                                      dtype=query.dtype,
-                                      device=query.device)
+        ] and not ascend_config.new_chunked:
+            attn_output_torch = torch.empty(num_tokens,
+                                            self.num_heads * self.v_head_dim,
+                                            dtype=query.dtype,
+                                            device=query.device)
             # current requests is chunked in prefill, disable flash attention with chunked prefill
             vanilla_chunked_prefill_mla(
```
Comment (ganyi1996ppo) [inline vllm_ascend/attention/mla_v1.py:759]: Please remove `vanilla_chunked_prefill_mla` and `flash_attention` path, we no longer need them

File: vllm_ascend/attention/mla_v1.py [hunk 5]
```diff
@@ -609,18 +772,47 @@ def _forward_prefill(
                 scale=self.scale,
                 alibi_slopes=None,
                 causal=True)
+        elif attn_metadata.attn_state in [
```
Comment (ganyi1996ppo) [inline vllm_ascend/attention/mla_v1.py:775]: We don't need `attn_state` any more you can just ignore all the attn_state, and just execute ring attention in forward_prefill by default.

File: vllm_ascend/attention/mla_v1.py [hunk 6]
```diff
@@ -638,6 +830,11 @@ def _forward_prefill(
             )
         attn_output = attn_output.reshape(
             [num_tokens, self.num_heads * self.v_head_dim])
+        if attn_metadata.attn_state in [
```
Comment (ganyi1996ppo) [inline vllm_ascend/attention/mla_v1.py:833]: ditto

File: tests/singlecard/test_chunked.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+import torch
+from vllm import LLM, SamplingParams
+
+MODELS = ["deepseek-ai/DeepSeek-V2-Lite"]
+
+
+@pytest.mark.skipif(os.getenv("VLLM_USE_V1") == "0",
+                    reason="new chunked only support on v1")
+@pytest.mark.parametrize("model", MODELS)
+@pytest.mark.parametrize("max_tokens", [1])
```
Comment (MengqingCao) [inline tests/singlecard/test_chunked.py:35]: please set `max_tokens` greater to make this test meaningful

File: tests/singlecard/test_chunked.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        logprobs_chunked = output_chunked.outputs[0].logprobs
+        del vllm_model
+        torch.npu.empty_cache()
+
+        vllm_model = LLM(model,
+                         enforce_eager=True,
+                         additional_config={
+                             'ascend_scheduler_config': {
+                                 'enabled': True
+                             },
```
Comment (MengqingCao) [inline tests/singlecard/test_chunked.py:65]: I think you should enable `chunked_prefill_for_mla` instead of `ascend_scheduler_config` ?

File: vllm_ascend/worker/model_runner_v1.py [hunk 9]
```diff
@@ -134,7 +134,11 @@ def __init__(self, vllm_config: VllmConfig, device: torch.device):
         self.lora_config = vllm_config.lora_config
         self.scheduler_config = vllm_config.scheduler_config
         self.speculative_config = vllm_config.speculative_config
-        self.chunked_prefill_enabled = vllm_config.scheduler_config.chunked_prefill_enabled
+        ascend_config = get_ascend_config()
+        if ascend_config.ascend_scheduler_config.enabled:
+            self.chunked_prefill_enabled = self.scheduler_config.chunked_prefill_enabled
+        else:
+            self.chunked_prefill_enabled = True
```
Comment (wangxiyuan) [inline vllm_ascend/worker/model_runner_v1.py:141]: this logic is correct. When using V1 with vllm scheduler, the `chunked_prefill_enabled` should be True. Otherwise, if using ascend scheduler, the value depends on the scheduler config.

**Review by wangxiyuan — APPROVED** [review]

File: docs/source/user_guide/additional_config.md [hunk 10]
```diff
@@ -31,6 +31,7 @@ The following table lists the additional configuration options available in vLLM
 | `expert_tensor_parallel_size` | str | `0`  | Expert tensor parallel size the model to use.                                                 |
 | `refresh`                     | bool | `false` | Whether to refresh global ascend config content. This value is usually used by rlhf case.     |
 | `expert_map_path`             | str | None | When using expert load balancing for the MOE model, an expert map path needs to be passed in. |
+| `chunked_prefill_for_mla` | bool | `False` | Whether to enable the fused operator-like chunked_prefill. |
```
Comment (wangxiyuan) [inline docs/source/user_guide/additional_config.md:34]: this config is only used temporary. we'll remove this value once pta is upgraded. And currently set this value to True will lead known error.

File: tests/singlecard/test_chunked.py (hunk 8 above)
Comment (wangxiyuan) [reply tests/singlecard/test_chunked.py:65]: @MengqingCao this is used for compare the result between  chunked prefill enabled or not. L52 enabled chunked prefill, and here disable by using ascend scheduler. So the test is fine.

File: vllm_ascend/attention/mla_v1.py (hunk 5 above)
Comment (wangxiyuan) [reply vllm_ascend/attention/mla_v1.py:775]: we'll remove it in the future once pta is upgraded.

**Review by wangxiyuan — COMMENTED** [review]
don't forget to remove `chunked_prefill_for_mla` once pta is upgraded.

**Comment by wangxiyuan** [comment]
@ganyi1996ppo  @Yikun @jianzs I talked with @fems14 offline. This change need more work to do base on new version of pta. She will update in the future. Currently, this change is backward compatible, so I merged it to unblock the feture. If you have any question, feel free to keep reviewing. The follow-up PR is welcome. Thanks.
